### PR TITLE
[niconico] Improvement niconico douga from another approach. (Support encrypted movies and old server source movies)

### PR DESCRIPTION
**Pull request by kurumigi**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/pukkandan/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of some part of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of rest of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Main part of this PR was pull requested two times by @tsukumijima (https://github.com/ytdl-org/youtube-dl/pull/23824 / https://github.com/blackjack4494/youtube-dlc/pull/55).

> niconico (ニコニコ動画), which is not known outside Japan, is the largest video sharing site in Japan.
> 
> niconico has begun using HLS to encrypt the distribution of some official videos, such as anime, on March 5, 2019.
> As a result, some official videos can no longer be downloaded with the previously used Extractor.
> (Examples of videos that can be downloaded as before: https://www.nicovideo.jp/watch/sm32868846
> (Example of video that cannot be downloaded due to HLS encryption: https://www.nicovideo.jp/watch/so36189921)
> 
> This pull request includes an improvement to allow you to download niconico's encrypted HLS videos.
> 
> In niconico, if you do not send "heartbeat" once every 60 seconds, the download session will be closed and you will get a 403 error.
> This pull request also includes a change to allow "heartbeat" to be sent once every 60 seconds during the download, and to allow long videos to be downloaded to the end.
> 
> We have also made changes so that you can obtain higher quality thumbnails and original titles.
> (Sentence may be a bit strange because we use translation)
> 

Additonaly, I write some code for downloading movie from SMILEVIDEO, niconico douga old distribution system. Niconico douga changed its distribution system from SMILEVIDEO to DMC (Dwango Media Cluster) in 2016. But SMILEVIDEO was stored source movies until 2018-06-28, and we can access this movies even now.

### CODE REVIEW

**Comment by pukkandan** [comment]
I didn't go through the code in-depth, but these are my initial thoughts:

1. Fix the conflicts - I am not knowledgeable in niconico, so it's totally up to you to resolve conflicts
2. Rebase the branch on yt-dlp and get rid of the merge commits from youtube-dlc. Having these unnecessary commits makes both review and conflict resolution tedious
3. From what I understand, the code is similar to https://github.com/animelover1984/youtube-dl, but without the live video extractor (and hence no need for websockets), correct?
4. The heartbeat logic should be moved to either a separate downloader similar to `YoutubeLiveChatReplayFD` or moved inside the extractor itself. (I can help with this)

**Comment by tsukumijima** [comment]
> From what I understand, the code is similar to https://github.com/animelover1984/youtube-dl, but without the live video extractor (and hence no need for websockets), correct?

Yes, it's correct. I'd like to support NicoLive in the future, but the drawback is that I need a WebSocket module. There seems to be a restriction that youtube-dl should not use external libraries, so I can't find a good solution.

> The heartbeat logic should be moved to either a separate downloader similar to `YoutubeLiveChatReplayFD` or moved inside the extractor itself. (I can help with this)

Heartbeat logic is an improvement over the code proposed by others, but I don't think the current implementation is good either.
I'm not familiar with Python, so I don't know how to solve it.

The sentence may be unnatural because it is machine translated from Japanese.

**Comment by pukkandan** [comment]
> I'm not familiar with Python, so I don't know how to solve it.

And, I am not familiar with niconico. So, unless someone who is familiar with it is willing to help, this PR probably won't go anywhere. Maybe @kurumigi or @bbepis would be interested?

>  I'd like to support NicoLive in the future, but the drawback is that I need a WebSocket module

Yes, youtube-dl has that requirement. But yt-dlp does not. However, if the dependency is not found, the import should fail gracefully. What I mean is, the `ImportError` should be handled silently and when the user tries to use a functionality that requires the dependancy, yt-dlp must give error message stating that an additional module is needed. You can see how I handled mutagen requirement for example:

https://github.com/pukkandan/yt-dlp/blob/b60419c51aa3eb9872e278e526cc5e62bf484462/youtube_dlc/postprocessor/embedthumbnail.py#L11-L15
https://github.com/pukkandan/yt-dlp/blob/b60419c51aa3eb9872e278e526cc5e62bf484462/youtube_dlc/postprocessor/embedthumbnail.py#L155-L156

**Comment by pukkandan** [comment]
@kurumigi Don't mind the linter. It is an issue from my end

**Comment by kurumigi** [comment]
> Fix the conflicts - I am not knowledgeable in niconico, so it's totally up to you to resolve conflicts
> Rebase the branch on yt-dlp and get rid of the merge commits from youtube-dlc. Having these unnecessary commits makes both review and conflict resolution tedious

Oops.
I pushd it now.

> From what I understand, the code is similar to https://github.com/animelover1984/youtube-dl, but without the live video extractor (and hence no need for websockets), correct?

Yes. I think only download of niconico douga (niconico's non-live service)..

File: youtube_dlc/extractor/niconico.py [hunk 1]
```diff
@@ -198,6 +196,46 @@ def yesno(boolean):
...
+
+        # hls (encryption)
+        if 'encryption' in api_data['video']['dmcInfo']:
+            session_api_http_parameters = {
+                'parameters': {
+                    'hls_parameters': {
+                        'encryption': {
+                            'hls_encryption_v1': {
+                                'encrypted_key': api_data['video']['dmcInfo']['encryption']['hls_encryption_v1']['encrypted_key'],
+                                'key_uri': api_data['video']['dmcInfo']['encryption']['hls_encryption_v1']['key_uri']
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:218]: Are these fields guaranteed to be present? Otherwise use try_get. Same issue in other places in code

File: youtube_dlc/extractor/niconico.py [hunk 2]
```diff
@@ -284,80 +329,79 @@ def _real_extract(self, url):
...
+            extension = determine_ext(video_real_url)
+
+        # Economy mode movie is not source movie.
+        # Similarly, if movie file size is unstable, old server movie is not source movie.
+        if not is_economy and int(get_video_info_xml('size_high')) > 1:
+            formats.append({
                 'url': video_real_url,
+                'format_id': 'smile',
+                'format_note': 'Source movie file in old server (SMILEVIDEO)',
+                'preference': -2,  # I think demand for source movie is not high.
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:389]: Please explain. This should probably be `source_preference`

File: youtube_dlc/extractor/niconico.py [hunk 3]
```diff
@@ -267,6 +304,14 @@ def yesno(boolean):
             'vbr': float_or_none(video_quality.get('bitrate'), 1000),
             'height': resolution.get('height'),
             'width': resolution.get('width'),
+            'quality': -2 if 'low' in format_id else -1,  # Default quality value is -1
+            'heartbeat_url': heartbeat_url,
+            'heartbeat_data': heartbeat_data,
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:309]: How does this affect `--write-info-json`? Can you give me an example of an output JSON?

Also test if `--load-info-json` is able to restart/continue the download

File: youtube_dlc/downloader/common.py [hunk 4]
```diff
@@ -372,7 +376,38 @@ def download(self, filename, info_dict, subtitle=False):
...
+
+            heartbeat()
+
+        try:
+            return self.real_download(filename, info_dict), True
+        finally:
+            if heartbeat_lock:
+                with heartbeat_lock:
+                    timer[0].cancel()
+                    download_complete = True
```
Comment (pukkandan) [inline youtube_dlc/downloader/common.py:410]: This code should be moved to a seperate downloader. I will look into how feasible it is to do. If it is too hard, we will use the code as-is.

Can you let me know what protocols videos from niconico can be?

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:389]: **Summary: Most of movie files in SMILEVIDEO (old server) before 2018-06-28 are SOURCE files. But, not recommend for obsolete encoding and lack of movie file information.**

According to official and unofficial technical documents written by Japanese.

Until 2018-06-28, under the following conditions SMILEVIDEO preserved uploaded movies without change.

- Movie encoding is swf, flv (H.263 / VP6), or mp4 (H.264).
- Movie size is below 100MB (uploaded by paid member) or 40MB (free member).
- Etc. Etc. (Japanese information : http://nicowiki.com/?cmd=backup&page=%E3%82%A8%E3%83%B3%E3%82%B3%E3%83%BC%E3%83%89%E8%A8%AD%E5%AE%9A&age=86#fe0ad87d )

The other cases, uploaded movies were re-encoded to mp4 (H.264).

However, SMILEVIDEO is deprecated system. So, download SMILEVIDEO movie file have several problems.

- Movie encoding : Now, swf and flv are obsolete. DMC (new server) always re-encoded them to mp4 (H.264).
- Reducing API : Now, information of SMILEVIDEO movie file from API are URL, extension and filesize only. API for DMC provided resolution and bitrate, mandatory information to detect download movie file.

Because I down 'preference', not 'source_preference'.

Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py:389]: ok, makes sense

File: youtube_dlc/extractor/niconico.py (hunk 3 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:309]: Test failed...

I have no idea now. Let me investigate about it.

Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py:309]: Probably because the heartbeat URL needs to be generated again. In that case, the relevant code must be moved to the downloader. The IE is not invoked when using load-info-json

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (bbepis) [reply youtube_dlc/extractor/niconico.py:389]: @kurumigi this is wrong. `smile_high` source files are the most important download from niconico because for the time period you specify (which might actually be around november 2016, i don't remember what my notes said), the smile_high videos are the highest quality for both video *and* audio.

The DMC streams are always worse in this regard, with degraded audio and reencoded video (which is usually downscaled to the nearest 16:9 standard resolution). The fact that some files use outdated codecs doesn't matter; any modern video player (VLC, MPC, MPV) can play them without issue. Since youtube-dl is an archival tool, the highest quality version should be preferred regardless. If you need certain codecs you can reencode them yourself, and you will still end up with a better quality video than the DMC versions because niconico uses very low quality settings for conversion.

There is no reason why they shouldn't be preferred (for videos uploaded before 2018 or 2016, need to double check). There are plenty of videos with 320kbit audio being reencoded to 192kbit in the DMC streams which is horrible. I can give plenty of examples of videos where quality is degraded from DMC conversion

File: youtube_dlc/downloader/common.py (hunk 4 above)
Comment (bbepis) [reply youtube_dlc/downloader/common.py:410]: Protocols are direct download for `smile`, and HLS for the DMC streams

**Comment by bbepis** [comment]
> And, I am not familiar with niconico. So, unless someone who is familiar with it is willing to help, this PR probably won't go anywhere. Maybe @kurumigi or @bbepis would be interested?

I have no idea how to do a clean job of this; I'm not familiar with python enough to do concurrency (at least without googling everything). There might be a clean way with async code but you don't wish to have the version requirement for that.

I do strongly feel that https://github.com/pukkandan/yt-dlp/pull/49#discussion_r569817286 should be marked as unresolved though, since it causes lower quality videos to be downloaded in a significant amount of videos

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py:389]: @kurumigi
`quality` is used to mark that a video is better than ALL videos of a higher `quality`
`source_quality` is used to indicate that the format has a source that is prefered over other sources (this does not significantly affect the actual format selection unless explicitly specified by the user with `-S source`)

Like I said before, I don't know anything about niconico. But if @bbepis is correct, the quality should not be set lower and instead `source_quality` should be set lower (since less metadata is available in the older source and therefore, the newer source is preferred if everything else is equal)

All this is assuming that the video codec and resolution (or bitrate) of the formats is known. If these cannot be determined, the situation is more complicated.

Please give me an example output of `-list-formats` of a video with both sources so that I can better understand the situation

**Comment by pukkandan** [comment]
> I have no idea how to do a clean job of this; I'm not familiar with python enough to do concurrency (at least without googling everything).

The current code is mostly good enough. Once https://github.com/pukkandan/yt-dlp/pull/49/files#discussion_r569649508 is resolved, I should be able to move the heartbeat code to a seperate downloader myself

> There might be a clean way with async code but you don't wish to have the version requirement for that.

See above comment https://github.com/pukkandan/yt-dlp/pull/49/files#issuecomment-771771482

> However, if the dependency is not found, the import should fail gracefully. What I mean is, the ImportError should be handled silently and when the user tries to use a functionality that requires the dependancy, yt-dlp must give error message stating that an additional module is needed. 

As long as the the async code can be made to fail gracefully in python2.6+ (and therefore not affect the use of other extractors), It can be used

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (bbepis) [reply youtube_dlc/extractor/niconico.py:389]: @pukkandan The main thing to grasp is that there are two distinct sources available for videos; `DMC` and `smile`.

Originally there was only the `smile` format; it would come in two variants, either `smile_high` or `smile_low` depending on if the website is currently under economy hours. This was the source video that was uploaded to NND, because niconico would cap uploads at 100MB

However sometime during 2016 (or 2018 according to that wiki linked), they introduced a new video serving system named `DMC` that would serve videos with HLS and standardized bitrates / resolutions. During this time they also removed the 100MB cap because everything would get reencoded regardless.

This is why the date is important; for videos uploaded before DMC was introduced, `smile_high` was the highest quality because it was the *original* video that was uploaded, and DMC sources are retroactively encoded from that. After that date, `smile_high` isn't necessarily the highest quality anymore because it's just a reencode designed to mimic the old limits, so the regular format selection rules apply to those.

I wrote something [more indepth in a reddit comment](https://www.reddit.com/r/youtubedl/comments/kluzux/i_have_a_fork_of_youtubedl_with_a_lot_of_fixes/gju6wob/?utm_source=reddit&utm_medium=web2x&context=3) but it focuses more on the history of the site

[Here is my logic](https://github.com/animelover1984/youtube-dl/blob/master/youtube_dl/extractor/niconico.py#L436-L491) that I use for `smile` format extraction, feel free to copy it outright.

I make use of ffprobe on the `smile` video to determine pretty much all metadata about the video because as of december 2020 the niconico API no longer contains any video info on the smile format for some reason.

Here is an example of the format selection that my fork gives for https://www.nicovideo.jp/watch/sm23133906:


image

Comment (bbepis) [reply youtube_dlc/extractor/niconico.py:389]: Here are some format selection results for SWF and FLV formats:
image

image

The API does not even tell you the format of the video, so ffprobe is required to even determine if it's an mp4, flv or swf file.

As I've said before if the formats are not suitable for the end user, they can either set `-f` to only grab the mp4/DMC variants, or use the remux post-processor to remux it to mkv (where it can then be played in any decent video player)

Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py:389]: Is there always only 1 smile format? Will it always be in the highest resolution?

Comment (bbepis) [reply youtube_dlc/extractor/niconico.py:389]: > Is there always only 1 smile format? Will it always be in the highest resolution?

`smile_high` is the higher quality, compared to `smile_low`. `smile_high` will only show if the user has a premium account, or it will show for free users of the site during offpeak hours (around 12 AM JST - 10 AM JST, they might have changed it).

From what I understand it's possible to show both formats in youtube-dl (assuming the user has access to the higher quality version), however I haven't figured out how to do it. Mainly because there wouldn't be any practical benefit, as the default version is the highest quality smile version that is available.

Comment (bbepis) [reply youtube_dlc/extractor/niconico.py:389]: Here's an example of a video uploaded in 2020 (after DMC was introduced) where the highest quality DMC video is better than `smile_high`:

image

File: youtube_dlc/extractor/niconico.py (hunk 1 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py]: Fix it with other similar parts for treatment more carefully.

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:387]: Hmm... Using ffprobe is an interesting idea.

Apart fromt hat,  API return "filesize is 1 byte" like this, the movie no longer stored in SMILEVIDEO.
This matter happens only in movies uploaded after 2018-06-28, it's one of the reasons I think "deadline" is 2018-06-28.

File: youtube_dlc/extractor/niconico.py (hunk 1 above)
Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py]: I'm marking this as resolved for now. I'll review again once the other issues are resolved

File: youtube_dlc/downloader/niconico.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .http import HttpFD
+from .external import FFmpegFD
+from ..extractor.niconico import NiconicoIE
+
```
Comment (pukkandan) [inline youtube_dlc/downloader/niconico.py:7]: It might be helpful to create a base class `NiconicoDmcFD` and derive these 2 classes from it. Like this:

    class NiconicoDmcHttpFD(NiconicoDmcFD, HttpFD):

It's just a suggestion. You don't need to do it if you think having separate classes is easier

Comment (kurumigi) [reply youtube_dlc/downloader/niconico.py:11]: Wow!

Comment (pukkandan) [reply youtube_dlc/downloader/niconico.py:11]: > Wow!

?

File: youtube_dlc/extractor/niconico.py (hunk 3 above)
Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py:313]: Upon further thought, this is not necessary (though still nice to have). Even for youtube extractor, the URL needs to be regenerated if it has expired. Yet, the code for it is in the extractor and not the downloader.

So don't worry about this too much. If needed, we can fix this separately in the future. Once the issue with format quality is resolved, I will merge the PR

File: youtube_dlc/downloader/niconico.py (hunk 5 above)
Comment (kurumigi) [reply youtube_dlc/downloader/niconico.py:11]: Oh, sorry. I have not-good image about multiple inheritence. But, python's it is very elegant, I surprised.

File: youtube_dlc/extractor/niconico.py (hunk 3 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py]: I managed to fix. I tested http and hls, passed both. Resume worked fine on http.

Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py]: great. I'll check it out

**Review by pukkandan — CHANGES_REQUESTED** [review]
@kurumigi I have reviewed the code so far and made some changes. Since I don't have write permission in your branch, I made a new branch with the changes. Please have a look https://github.com/pukkandan/yt-dlp/commits/nico

**Comment by kurumigi** [comment]
@pukkandan Thanks for adjustments! I tested using "--external_downloader ffmpeg", downloaded no problem.

**Review by pukkandan — APPROVED** [review]

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py:419]: now only this issue left

**Comment by tsukumijima** [comment]
The heartbeat related processing is nicely coded and nice. Thank you.

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:419]: I re-check official and unofficial technical documents written by Japanese, I understand **major** specification changes of niconico douga are two times.

- **Movies uploaded before 2016-12-08**:
    **Maximum Movie filesize was 100MB**.
    SMILEVIDEO preserved acceptable encode (wrote in previous comment) movies, and re-encoded unacceptable movies to H.264.
    After 2018-04-10, DMC re-encoded all movies from SMILEVIDEO. (Official Source: http://blog.nicovideo.jp/niconews/69409.html (Japanese))
    
    **Conclusion : SMILEVIDEO's movie are always superior to DMC's one.**

- **Movies uploaded between 2016-12-08 and 2018-06-28**:
    **Maximum Movie filesize was upped to 1.5GB.**
    SMILEVIDEO preserved acceptable encode movies **below 100MB**, and re-encoded unacceptable movies to H.264 **below 100MB**. (Archive of Official Source: https://www.nicovideo.jp/watch/sm33168111 (Japanese / Movie))
    DMC re-encoded all movies immediately.
    
    **Conclusion : SMILEVIDEO acceptable movies are superior to DMC's one. Other movies depends re-encoded movie size.**
    
- **Movies uploaded after 2018-06-28**:
    SMILEVIDEO deactivated. No longer stored movies. (Official Source: https://twitter.com/nico_nico_talk/status/1004290141530173440 (Japanese))
    DMC re-encoded all movies immediately.
    
    **Conclusion : DMC's movie only.**

Hmmmm....

Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:419]: @bbepis

> Here is my logic that I use for smile format extraction, feel free to copy it outright.

 I will merge this code. Thank you for permission to copy code.

**Comment by pukkandan** [comment]
Let me know when this is ready for final review

**Comment by pukkandan** [comment]
Don't worry about the merge conflict. It is caused because I have changed how nested downloaders (like niconico) should behave.
**I will resolve it** once the code is complete from your end.

File: youtube_dlc/postprocessor/ffmpeg.py [hunk 6]
```diff
@@ -225,6 +225,11 @@ def get_metadata_object(self, path, opts=[]):
             self._downloader.to_screen('[debug] ffprobe command line: %s' % shell_quote(cmd))
         p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
         stdout, stderr = p.communicate()
+        if self._downloader.params.get('verbose', False):
+            self._downloader.to_screen('[debug] ffprobe output: %s' % stdout.decode('utf-8', 'replace'))
+        if self._downloader.params.get('verbose', False):
+            self._downloader.to_screen('[debug] ffprobe stderr: %s' % stderr.decode('utf-8', 'replace'))
+
```
Comment (pukkandan) [inline youtube_dlc/postprocessor/ffmpeg.py:232]: this is unrelated to this PR

Comment (pukkandan) [inline youtube_dlc/postprocessor/ffmpeg.py:232]: Unrelated to this PR

File: youtube_dlc/extractor/niconico.py [hunk 7]
```diff
@@ -284,80 +360,132 @@ def _real_extract(self, url):
...
+            or 'mp4' if 'mp4' in metadata['format']['format_name'] else metadata['format']['format_name']
+            or determine_ext(video_real_url)
+        )
+
+        timestamp = (
+            parse_iso8601(get_video_info_web('first_retrieve'))
+            or unified_timestamp(get_video_info_web('postedDateTime'))
+        )
+
+        # According to my personal research,
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:449]: Remove this line or give source

File: youtube_dlc/extractor/niconico.py [hunk 8]
```diff
@@ -284,80 +360,132 @@ def _real_extract(self, url):
...
+        # 'creation_time' tag on video stream of re-encoded SMILEVIDEO mp4 files are '1970-01-01T00:00:00.000000Z'.
+        metadata_timestamp = (
+            parse_iso8601(try_get(v_stream, lambda x: x['tags']['creation_time']))
+            or timestamp
+        )
+
+        # According to compconf and my personal research, smile videos from pre-2017 are always better quality than their DMC counterparts
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:456]:
```suggestion
        # According to compconf, smile videos from pre-2017 are always better quality than their DMC counterparts
```

or better, give source

File: youtube_dlc/extractor/niconico.py [hunk 9]
```diff
@@ -284,80 +360,132 @@ def _real_extract(self, url):
...
-                    'url': video_real_url,
-                    'ext': 'mp4',
-                    'format_id': _format_id_from_url(video_real_url),
-                }]
-
-            def get_video_info(items):
-                return dict_get(api_data['video'], items)
+                'container': extension,
+                'vcodec': dict_get(v_stream, 'codec_name'),
+                'acodec': dict_get(a_stream, 'codec_name'),
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:470]:
```suggestion
                'vcodec': v_stream.get('codec_name'),
                'acodec': a_stream.get('codec_name'),
```

No need to over-complicate since we are only looking for a single key

File: youtube_dlc/extractor/niconico.py [hunk 10]
```diff
@@ -284,80 +360,132 @@ def _real_extract(self, url):
...
+                # Some complex swf files doesn't have total bit rate metadata (e.g. nm6049209)
+                'tbr': int_or_none(metadata['format'].get('bit_rate', None), scale=1000),
+                'vbr': int_or_none(v_stream.get('bit_rate', None), scale=1000),
+                'abr': int_or_none(a_stream.get('bit_rate', None), scale=1000),
+                'height': int_or_none(v_stream.get('height', None)),
+                'width': int_or_none(v_stream.get('width', None)),
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:476]:
```suggestion
                'tbr': int_or_none(metadata['format'].get('bit_rate'), scale=1000),
                'vbr': int_or_none(v_stream.get('bit_rate'), scale=1000),
                'abr': int_or_none(a_stream.get('bit_rate'), scale=1000),
                'height': int_or_none(v_stream.get('height')),
                'width': int_or_none(v_stream.get('width')),
```

`None` is the default for `.get`

**Review by pukkandan — COMMENTED** [review]
Just a few small things. Leaving review for my future reference, I will fix these

File: youtube_dlc/extractor/niconico.py (hunk 2 above)
Comment (pukkandan) [reply youtube_dlc/extractor/niconico.py]: @bbpeis Could you review the metadata extraction. I don't fully understand it

**Comment by kurumigi** [comment]
Sorry.
I forget to check on community members only movies.
Fix about them.

File: youtube_dlc/extractor/niconico.py [hunk 11]
```diff
@@ -386,8 +385,7 @@ def get_video_info_xml(items):
                 raise ExtractorError('The video is not found.',
                                      expected=True)
             elif error_code == 'COMMUNITY':
-                raise ExtractorError('The video is community members only.',
-                                     expected=True)
+                self.to_screen('%s: The video is community members only.' % video_id)
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:388]: Why `to_screen`? Can it still be downloaded?

File: youtube_dlc/extractor/niconico.py [hunk 12]
```diff
@@ -284,80 +359,127 @@ def _real_extract(self, url):
...
+        if is_economy:
+            self.report_warning('Site is currently in economy mode! You will only have access to lower quality streams')
+
+        # Invoking ffprobe to determine resolution
+        pp = FFmpegPostProcessor(self._downloader)
+        cookies = self._get_cookies('https://nicovideo.jp').output(header='', sep='; path=/; domain=nicovideo.jp;\n')
+
+        self.to_screen('%s: %s' % (video_id, 'Checking smile format with ffprobe'))
+
+        metadata = pp.get_metadata_object(video_real_url, ['-cookies', cookies])
```
Comment (pukkandan) [inline youtube_dlc/extractor/niconico.py:419]: `PostProcessorError` must be caught and raised as `ExtractorError`

**Review by pukkandan — APPROVED** [review]
Everything looks good.

Is this ready for merge, or are you still making changes?

File: youtube_dlc/extractor/niconico.py (hunk 11 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:388]: Yes. Community members can't download community members only movies.
But, raise ExtractorError another place for other users.

File: youtube_dlc/extractor/niconico.py (hunk 12 above)
Comment (kurumigi) [reply youtube_dlc/extractor/niconico.py:419]: OK.

**Comment by pukkandan** [comment]
Merging it now. If you have any further improvements, make a new PR
